dict.js

```diff
@@ -46,74 +35,123 @@ Dict.prototype.assertString = function (key) {
     this.assertString(key);
-    var mangled = mangle(key);
-    if (mangled in this.store) {
-        return this.store[mangled];
-    } else if (arguments.length > 1) {
-        return defaultValue;
-    } else {
-        return this.getDefault(key);
+    if (key === "__proto__") {
+        if (this.hasOwnProperty("$__proto__")) {
+            return this["$__proto__"];
+        } else if (arguments.length > 1) {
+            return defaultValue;
+        } else {
+            return this.getDefault(key);
+        }
+    }
+    else {
+        if (key in this.store) {
+            return this.store[key];
+        } else if (arguments.length > 1) {
+            return defaultValue;
+        } else {
+            return this.getDefault(key);
+        }
     }
 };
 
 Dict.prototype.set = function (key, value) {
     this.assertString(key);
-    var mangled = mangle(key);
-    if (mangled in this.store) { // update
-        if (this.dispatchesMapChanges) {
-            this.dispatchBeforeMapChange(key, this.store[mangled]);
-        }
-        this.store[mangled] = value;
-        if (this.dispatchesMapChanges) {
-            this.dispatchMapChange(key, value);
-        }
-        return false;
-    } else { // create
-        if (this.dispatchesMapChanges) {
-            this.dispatchBeforeMapChange(key, undefined);
+    if (key === "__proto__") {
+        if (this.hasOwnProperty("$__proto__")) { // update
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, this["$__proto__"]);
+            }
+            this["$__proto__"] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return false;
+        } else { // create
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, undefined);
+            }
+            this.length++;
+            this["$__proto__"] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return true;
         }
-        this.length++;
-        this.store[mangled] = value;
-        if (this.dispatchesMapChanges) {
-            this.dispatchMapChange(key, value);
+    }
+    else {
+        if (key in this.store) { // update
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, this.store[key]);
+            }
+            this.store[key] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return false;
+        } else { // create
+            if (this.dispatchesMapChanges) {
+                this.dispatchBeforeMapChange(key, undefined);
+            }
+            this.length++;
+            this.store[key] = value;
+            if (this.dispatchesMapChanges) {
+                this.dispatchMapChange(key, value);
+            }
+            return true;
```

Comment:
Yes, considered that but worried it would slowdown Dict for most use cases, storing "__proto__" is most likely exotic statically speaking


montagestudio.com
 	
Benoit Marchant
CEO & Founder
1290 Oakmead Parkway,y Ste 111
Sunnyvale, CA 94085
(1) 408.621.4874 mobile
Profile | @benoitmarchant

> On May 4, 2015, at 19:37, Kris Kowal <notifications@github.com> wrote:
> 
> In dict.js:
> 
> > +            }
> > +            this.store[key] = value;
> > +            if (this.dispatchesMapChanges) {
> > +                this.dispatchMapChange(key, value);
> > +            }
> > +            return false;
> > +        } else { // create
> > +            if (this.dispatchesMapChanges) {
> > +                this.dispatchBeforeMapChange(key, undefined);
> > +            }
> > +            this.length++;
> > +            this.store[key] = value;
> > +            if (this.dispatchesMapChanges) {
> > +                this.dispatchMapChange(key, value);
> > +            }
> > +            return true;
> Might be less duplicate code if you do the special logic for "proto" immediately around the this.store manipulation lines.
> 
> —
> Reply to this email directly or view it on GitHub.
>